File: Algorithms/Bit Manipulation/137. Single Number II.py
#hashmap
class Solution(object):
	def singleNumber(self, nums):
		D = dict()
		for i in nums:
			D[i] = D.setdefault(i, 0) + 1
		for i in D:
			if(D[i] == 1):
				return i

#bit manipulation
#00 > 10 > 01 > 00
class Solution(object):
	def singleNumber(self, nums):	
		ones, twos = 0, 0
		for i in nums:
			ones = (ones ^ i) & ~twos
			twos = (twos ^ i) & ~ones
		return ones

File: Algorithms/Bit Manipulation/137. Single Number II.cpp
/////  num   one   two
/////   1     0     0
/////   1     1     0
/////   1     0     1
/////   1     0     0
/////   1     1     0
/////   1     0     1

/////  num   one   two
/////   0     0     0
/////   0     0     0

/////  num   one   two
/////   0     1     0
/////   0     1     0

/////  num   one   two
/////   0     0     1
/////   0     0     1

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int one = 0, two = 0;
        for(int& num : nums)
        {
            one = (one ^ num) & ~two;
            two = (two ^ num) & ~one;
        }
        return one;
    }
};
